// @eueno/lib-node/src/config.js
module.exports = {
  announceList: ['wss://tracker.eueno.io', 'https://tracker.eueno.io/announce'],
  api: {
    requestUrlUpload: '/sign-service/sdk/v1/auth-upload',
    shareTo: '/share-service/v1/api/shares',
    uploadKeyEncrypt: '/project-service/v1/api/files/update-encrypt-key',
    login: '/api/v1/users/login',
    requestUrlUploadList: '/sign-service/sdk/v1/auth-upload-lists',
    getFileList: '/project-service/v1/api/files/items-in-project/:project_id?params',
    getFileDetail: '/project-service/v1/api/files/:fileId',
    deleteFile: '/project-service/v1/api/files/:fileId',
    getSharedFiles:
      '/share-service/v1/api/shares/share-with-me?page={page}&limit={limit}&projectId={projectId}',
    createProject: '/project-service/v1/api/projects',
    getPersonalProject: '/project-service/v1/api/projects/personal?page={page}&limit={limit}&sortBy={sortBy}',
    createFolder: '/project-service/v1/api/files/folder',
    createAccount: '/account-service/sdk/users/login',
  },
  ipfsGateway: 'https://node1-gateway-ipfs.eueno.io',
};
