// @eueno/lib-node/src/project.js
const request = require('request');
const { api } = require('./config');

module.exports = {
  callCreateProject: async (opts) => {
    return new Promise((resolve) => {
      request(
        {
          method: 'POST',
          url: `${opts.endpoint}${api.createProject}`,
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            name: opts.projectName,
            signData: opts.signData,
          }),
        },
        function(error, response) {
          if (error) throw new Error(error);
          resolve(JSON.parse(response.body));
        },
      );
    });
  },
  callGetPersonalProject: async (opts) => {
    return new Promise((resolve) => {
      request(
        {
          method: 'POST',
          url: `${opts.endpoint}${api.getPersonalProject}`
            .replace('{page}', opts.page)
            .replace('{limit}', opts.limit)
            .replace('{sortBy}', opts.sortBy),
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            signData: opts.signData,
          }),
        },
        function (error, response) {
          if (error) throw new Error(error);
          resolve(JSON.parse(response.body));
        }
      );
    });
  },
};
